cmd/argocd/commands/repo.go

```diff
@@ -60,24 +60,32 @@ func NewRepoAddCommand(clientOpts *argocdclient.ClientOptions) *cobra.Command {
 				}
 				repo.SSHPrivateKey = string(keyData)
 			}
+			// InsecureIgnoreHostKey is deprecated and only here for backwards compat
 			repo.InsecureIgnoreHostKey = insecureIgnoreHostKey
-			// First test the repo *without* username/password. This gives us a hint on whether this
-			// is a private repo.
-			// NOTE: it is important not to run git commands to test git credentials on the user's
-			// system since it may mess with their git credential store (e.g. osx keychain).
-			// See issue #315
-			err := git.TestRepo(repo.Repo, "", "", repo.SSHPrivateKey, repo.InsecureIgnoreHostKey)
-			if err != nil {
-				if yes, _ := git.IsSSHURL(repo.Repo); yes {
-					// If we failed using git SSH credentials, then the repo is automatically bad
-					log.Fatal(err)
-				}
-				// If we can't test the repo, it's probably private. Prompt for credentials and
-				// let the server test it.
-				repo.Username, repo.Password = cli.PromptCredentials(repo.Username, repo.Password)
-			}
+			repo.Insecure = insecureSkipServerVerification
+
 			conn, repoIf := argocdclient.NewClientOrDie(clientOpts).NewRepoClientOrDie()
 			defer util.Close(conn)
+
+			// If the user set a username, but didn't supply password via --password,
```

Comment:
Looks like we've slightly changed behavior here. The intent was to validate repo and if it is accessible don't even ask username/password. If repo is not accessible and URL looks like https then ask for username/password. 

Can we keep the same behavior but use `repoIf.ValidateAccess` instead of `git.TestRepo`?

Reply:
I think to keep previous behaviour, we would possibly have to send two requests to the server before adding the repository:

First request checks repository access, and on error, we have to evaluate why access was not possible. In case of inability to validate server cert, let the user know and bail out (user has then either to add custom cert, or specify `--insecure-skip-server-verification` for the next run). 

In case of required authorization (then HTTPS connection was successful), prompt the user for credentials and call `repoIf.ValidateAccess` again with the supplied credentials.

However, this would mean to change `repoif.ValidateAccess` to return some meaningful response instead of generic `error`, so that we can differentiate between the reasons for no access. But I think that's a good idea anyway.